lib/posix/pthread.c

```diff
@@ -36,23 +56,68 @@ static const struct pthread_attr init_pthread_attrs = {
 	.initialized = true,
 };
 
-static struct posix_thread posix_thread_pool[CONFIG_MAX_PTHREAD_COUNT];
-static struct k_spinlock pthread_pool_lock;
+/*
+ * We reserve the MSB to mark a pthread_t as initialized (from the
+ * perspective of the application). With a linear space, this means that
+ * the theoretical pthread_t range is [0,2147483647].
+ */
+BUILD_ASSERT(CONFIG_MAX_PTHREAD_COUNT < PTHREAD_OBJ_MASK_INIT,
+	     "CONFIG_MAX_PTHREAD_COUNT is too high");
 
-pthread_t pthread_self(void)
+static inline size_t posix_thread_to_offset(struct posix_thread *t)
+{
+	return t - posix_thread_pool;
+}
+
+static inline size_t get_posix_thread_idx(pthread_t pth)
 {
-	return (struct posix_thread *)
-		CONTAINER_OF(k_current_get(), struct posix_thread, thread)
-		- posix_thread_pool;
+	return mark_pthread_obj_uninitialized(pth);
 }
 
-struct posix_thread *to_posix_thread(pthread_t pthread)
+struct posix_thread *to_posix_thread(pthread_t pth)
 {
-	if (pthread >= CONFIG_MAX_PTHREAD_COUNT) {
+	k_spinlock_key_t key;
+	struct posix_thread *t;
+	bool actually_initialized;
+	size_t bit = get_posix_thread_idx(pth);
+
+	/* if the provided thread does not claim to be initialized, its invalid */
+	if (!is_pthread_obj_initialized(pth)) {
+		return NULL;
+	}
+
+	if (bit >= CONFIG_MAX_PTHREAD_COUNT) {
+		return NULL;
+	}
+
+	t = &posix_thread_pool[bit];
+
+	key = k_spin_lock(&q_lock);
+	/*
+	 * Denote a pthread as "initialized" (i.e. allocated) if it is not in ready_q.
+	 * This differs from other posix object allocation strategies because they use
+	 * a bitarray to indicate whether an object has been allocated.
+	 */
+	actually_initialized = t->qid != POSIX_THREAD_READY_Q;
```

Comment:
Seems like this is just doing a single load under the spin lock? Might be nice to allow for architectures for which this is an atomic operation (essentially all of them)?

Reply:
I'm not sure if we have an atomic load for `uint8_t`, and everywhere else that touches the queues uses the spinlock as well.

My main goal with this change was eliminating the remaining race condition.

There are likely lots of micro-optimizations to be made, but I would be happy to optimize later.

Reply:
yeah, good point. Get it working, then get it working fast.